src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -133,103 +169,67 @@ class InterceptedUnaryUnaryCall(_base_call.UnaryUnaryCall):
-            self._fire_pending_add_done_callbacks)
+            self._fire_or_add_pending_add_done_callbacks)
 
     def __del__(self):
         self.cancel()
 
-    # pylint: disable=too-many-arguments
-    async def _invoke(self, interceptors: Sequence[UnaryUnaryClientInterceptor],
-                      method: bytes, timeout: Optional[float],
-                      metadata: Optional[MetadataType],
-                      credentials: Optional[grpc.CallCredentials],
-                      wait_for_ready: Optional[bool], request: RequestType,
-                      request_serializer: SerializingFunction,
-                      response_deserializer: DeserializingFunction
-                     ) -> UnaryUnaryCall:
-        """Run the RPC call wrapped in interceptors"""
-
-        async def _run_interceptor(
-                interceptors: Iterator[UnaryUnaryClientInterceptor],
-                client_call_details: ClientCallDetails,
-                request: RequestType) -> _base_call.UnaryUnaryCall:
-
-            interceptor = next(interceptors, None)
-
-            if interceptor:
-                continuation = functools.partial(_run_interceptor, interceptors)
+    def _fire_or_add_pending_add_done_callbacks(self,
+                                                interceptors_task: asyncio.Task
+                                               ) -> None:
 
-                call_or_response = await interceptor.intercept_unary_unary(
-                    continuation, client_call_details, request)
-
-                if isinstance(call_or_response, _base_call.UnaryUnaryCall):
-                    return call_or_response
-                else:
-                    return UnaryUnaryCallResponse(call_or_response)
+        if not self._pending_add_done_callbacks:
+            return
 
-            else:
-                return UnaryUnaryCall(
-                    request, _timeout_to_deadline(client_call_details.timeout),
-                    client_call_details.metadata,
-                    client_call_details.credentials,
-                    client_call_details.wait_for_ready, self._channel,
-                    client_call_details.method, request_serializer,
-                    response_deserializer, self._loop)
+        fire = False
 
-        client_call_details = ClientCallDetails(method, timeout, metadata,
-                                                credentials, wait_for_ready)
-        return await _run_interceptor(iter(interceptors), client_call_details,
-                                      request)
+        try:
+            call = interceptors_task.result()
+            if call.done():
+                fire = True
+        except (AioRpcError, asyncio.CancelledError):
+            fire = True
 
-    def _fire_pending_add_done_callbacks(self,
-                                         unused_task: asyncio.Task) -> None:
         for callback in self._pending_add_done_callbacks:
-            callback(self)
+            if fire:
```

Comment:
The value of `fire` doesn't ever change inside of this loop, right? So we should be able to pull the conditional outside of the loop and get a performance boost.